apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,409 @@
+                    /*
+                    let busy = span
+                        .attributes
+                        .get(&opentelemetry::Key::from_static_str("busy_ns"))
+                        .unwrap();
+                    let busy_v = match busy {
+                        Value::I64(v) => v / 1_000_000,
+                        _ => panic!("value should be a signed integer"),
+                    };
+                    */
+                    tracing::debug!("query: {}", q);
+                    tracing::debug!("busy: {}", busy_v);
+
+                    let not_found = Value::String(Cow::from("not found"));
+                    let stats = ContextualizedStats {
+                        context: Some(StatsContext {
+                            client_name: span
+                                .attributes
+                                .get(&opentelemetry::Key::from_static_str("client_name"))
+                                .unwrap_or(&not_found)
+                                .to_string(),
+                            client_version: span
+                                .attributes
+                                .get(&opentelemetry::Key::from_static_str("client_version"))
+                                .unwrap_or(&not_found)
+                                .to_string(),
+                        }),
+                        query_latency_stats: Some(QueryLatencyStats {
+                            latency_count: vec![busy_v],
+                            request_count: 1,
+                            ..Default::default()
+                        }),
+                        ..Default::default()
+                    };
+                    let operation_name = span
+                        .attributes
+                        .get(&opentelemetry::Key::from_static_str("operation_name"));
+                    // XXX NEED TO NORMALISE THE QUERY
+                    let key = normalize(operation_name, &q.as_str());
+
+                    let msg = reporter
+                        .submit_stats(graph.clone(), key, stats)
+                        .await
+                        .map_err::<TraceError, _>(|e| e.to_string().into())?
+                        .into_inner()
+                        .message;
+                    tracing::info!("server response: {}", msg);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn normalize(op: Option<&opentelemetry::Value>, q: &str) -> String {
+    // If we don't have an operation name, no point normalizing
+    // it. Just return the unprocessed input.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => return q.to_string(),
+    };
+    let parser = Parser::new(q);
+    // compress *before* parsing to modify whitespaces/comments
+    let ast = parser.compress().parse();
+    tracing::debug!("ast:\n {:?}", ast);
+    // If we can't parse the query, we definitely can't normalize it, so
```

Comment:
We used to do that but that led to (a) tons of cardinality coming from broken input and (b) unsanitized operations getting sent to us. Instead we have special stats report keys for the error cases of "parse error" (invalid GraphQL syntax), "validation error" (operation doesn't work with the schema), and "unknown operation name" (the document is valid but the separately-provided operation name doesn't match an operation in the doc, or isn't provided and there are multiple ops there).

https://github.com/apollographql/apollo-server/blob/d1b19d07af6d660e18021d7c14b91e7c555b4d73/packages/apollo-server-core/src/plugin/usageReporting/plugin.ts#L639-L645

Reply:
This is WIP until the normalisation support in apollo-rs is available. I'll look at this again then.

Reply:
By the way I thought it might be helpful for me to be more explicit about what the three kinds of top-level errors mean, since it's a bit subtle.

A GraphQL request (over HTTP/JSON) consists of up to three fields:
- `query` (poorly named as it can also contain mutations etc): a required GraphQL document
- `operationName`: an optional operation name, used to select between multiple operations in `query` (and can also be specified even if there's only one, but it must match if provided)
- `variables`: the data that varies between different executions of the same operation

(we also have added `extensions` to that; the spec only specifies this for responses but we use it for stuff like automatic persisted queries.)

GraphQL requests are executed in the context of a `schema`.

- Whether or not something is a "parse error" is a function purely of `query` (and, I mean, the version of the graphql library/language in use of course), not of `schema`, `operationName`, or `variables`. It happens when the actual language in `query` fails to parse. (Servers can cache the result of parsing indefinitely.)
- Whether or not something is a "validation error" is a function of `query` and `schema`, but not of `operationName` or `variables`.  It happens when the properly-parsing document tries to do things that don't work with the given schema. It consists of a bunch of rules defined in [the validation section of the spec](http://spec.graphql.org/draft/#sec-Validation). (Servers can cache the result of validation for as long as the underlying schema does not change.)
- Whether or not something is an "unknown operation name" error is a function of `query` and `operationName`. According to how the spec is laid out (and how the reference implementation `graphql-js` works) checking for this is actually the first step of the execution phase, but we find it helpful to effectively separate it into its own page (for various reasons but largely because our usage reporting is based on the specific executed operation so we want to treat this kind of error as distinct from other execution errors).

The other kind of errors that can arise happen at execution time. Most of them are "field errors"; the one other kind of "request error" is if a variable value cannot be coerced to its declared type.